src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -540,64 +537,436 @@ void AtlasEngine::_drawCursor()
-        rect.bottom -= halfWidth;
+        box.left += halfWidth;
+        box.top += halfWidth;
+        box.right -= halfWidth;
+        box.bottom -= halfWidth;
         break;
     }
     case CursorType::Underscore:
     case CursorType::DoubleUnderscore:
-        rect.top = _r.cellSizeDIP.y - lineWidth;
+        box.top = box.bottom - lineWidth;
         break;
     default:
         break;
     }
 
-    _r.d2dRenderTarget->BeginDraw();
+    const auto brush = _brushWithColor(color);
+
     // We need to clip the area we draw in to ensure we don't
     // accidentally draw into any neighboring texture atlas tiles.
     _r.d2dRenderTarget->PushAxisAlignedClip(&clip, D2D1_ANTIALIAS_MODE_ALIASED);
-    _r.d2dRenderTarget->Clear();
+
+    if (clear)
+    {
+        _r.d2dRenderTarget->Clear();
+    }
 
     if (cursorType == CursorType::EmptyBox)
     {
-        _r.d2dRenderTarget->DrawRectangle(&rect, _r.brush.get(), lineWidth);
+        _r.d2dRenderTarget->DrawRectangle(&box, brush, lineWidth);
     }
     else
     {
-        _r.d2dRenderTarget->FillRectangle(&rect, _r.brush.get());
+        _r.d2dRenderTarget->FillRectangle(&box, brush);
     }
 
     if (cursorType == CursorType::DoubleUnderscore)
     {
-        rect.top -= 2.0f;
-        rect.bottom -= 2.0f;
-        _r.d2dRenderTarget->FillRectangle(&rect, _r.brush.get());
+        const auto offset = lineWidth * 2.0f;
+        box.top -= offset;
+        box.bottom -= offset;
+        _r.d2dRenderTarget->FillRectangle(&box, brush);
     }
 
     _r.d2dRenderTarget->PopAxisAlignedClip();
+}
+
+ID2D1Brush* AtlasEngine::_brushWithColor(u32 color)
+{
+    if (_r.brushColor != color)
+    {
+        const auto d2dColor = colorFromU32(color);
+        THROW_IF_FAILED(_r.d2dRenderTarget->CreateSolidColorBrush(&d2dColor, nullptr, _r.brush.put()));
+        _r.brushColor = color;
+    }
+    return _r.brush.get();
+}
+
+AtlasEngine::CachedGlyphLayout::operator bool() const noexcept
+{
+    return static_cast<bool>(textLayout);
+}
+
+void AtlasEngine::CachedGlyphLayout::reset() noexcept
+{
+    textLayout.reset();
+}
+
+void AtlasEngine::CachedGlyphLayout::applyScaling(ID2D1RenderTarget* d2dRenderTarget, D2D1_POINT_2F origin) const noexcept
+{
+    __assume(d2dRenderTarget != nullptr);
+
+    if (scalingRequired)
+    {
+        const D2D1_MATRIX_3X2_F transform{
+            scale.x,
+            0,
+            0,
+            scale.y,
+            (origin.x + halfSize.x) * (1.0f - scale.x),
+            (origin.y + halfSize.y) * (1.0f - scale.y),
+        };
+        d2dRenderTarget->SetTransform(&transform);
+    }
+}
+
+void AtlasEngine::CachedGlyphLayout::undoScaling(ID2D1RenderTarget* d2dRenderTarget) const noexcept
+{
+    __assume(d2dRenderTarget != nullptr);
+
+    if (scalingRequired)
+    {
+        static constexpr D2D1_MATRIX_3X2_F identity{ 1, 0, 0, 1, 0, 0 };
+        d2dRenderTarget->SetTransform(&identity);
+    }
+}
+
+void AtlasEngine::_d2dPresent()

```

Comment:
aside - does it almost make sense to just have `AtlasD2D.r.cpp` or something for all these `AtlasEngine::_d2d*` methods?

Reply:
I'm planning a larger reorganization of the code at a later point. I'm not 100% sure whether such a split will be helpful, since the `_d2d*` function do use a lot of common logic of the regular AtlasEngine code, but if we want to split it I'd do it then.